x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/transform/action/GetCheckpointAction.java

```diff
@@ -0,0 +1,140 @@
+    }
+
+    public static class Request extends ActionRequest implements IndicesRequest {
+
+        private final String[] indices;
+        private final IndicesOptions indicesOptions;
+
+        public Request(StreamInput in) throws IOException {
+            super(in);
+            indices = in.readStringArray();
+            indicesOptions = IndicesOptions.readIndicesOptions(in);
+        }
+
+        public Request(String[] indices, IndicesOptions indicesOptions) {
+            this.indices = indices;
+            this.indicesOptions = indicesOptions;
+        }
+
+        @Override
+        public ActionRequestValidationException validate() {
+            return null;
+        }
+
+        @Override
+        public String[] indices() {
+            return indices;
+        }
+
+        @Override
+        public IndicesOptions indicesOptions() {
+            return indicesOptions;
+        }
+
+        @Override
+        public boolean equals(Object obj) {
+            if (obj == this) {
+                return true;
+            }
+            if (obj == null || obj.getClass() != getClass()) {
+                return false;
+            }
+            Request that = (Request) obj;
+
+            return Arrays.equals(indices, that.indices) && Objects.equals(indicesOptions, that.indicesOptions);
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(Arrays.hashCode(indices), indicesOptions);
+        }
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
+            super.writeTo(out);
+            out.writeStringArrayNullable(indices);
+            indicesOptions.writeIndicesOptions(out);
+        }
+    }
+
+    public static class Response extends ActionResponse {
+
+        private final Map<String, long[]> checkpoints;
+
+        public Response(Map<String, long[]> checkpoints) {
+            this.checkpoints = checkpoints;
+        }
+
+        public Response(StreamInput in) throws IOException {
+            this.checkpoints = in.readOrderedMap(StreamInput::readString, StreamInput::readLongArray);
+        }
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
+            out.writeMap(getCheckpoints(), StreamOutput::writeString, StreamOutput::writeLongArray);
+        }
+
+        public Map<String, long[]> getCheckpoints() {
+            return Collections.unmodifiableMap(checkpoints);
+        }
+
+        @Override
+        public boolean equals(Object obj) {
+            if (obj == this) {
+                return true;
+            }
+            if (obj == null || obj.getClass() != getClass()) {
+                return false;
+            }
+            Response that = (Response) obj;
+
+            return this.checkpoints.size() == that.checkpoints.size()
+                && this.checkpoints.entrySet().stream().allMatch(e -> Arrays.equals(e.getValue(), that.checkpoints.get(e.getKey())));
```

Comment:
Why can't we use `Objects.equals` here?

Reply:
`Objects.equals` compares references but not the content:

```
    public void testEquals() {
        int[] arr1 = new int[10]; // initialized with 0 in all cells
        int[] arr2 = new int[10]; // initialized with 0 in all cells
        //assertTrue(Objects.equals(arr1, arr2)); // false
        assertTrue(Arrays.equals(arr1, arr2)); // true
    }
    ```